File: lua/day3/main.lua
local function getData(path)
  local data = {}
  for line in io.lines(path) do
    table.insert(data, line)
  end
  return data
end

local function getPriority(c)
  local ascii = c:byte()
  if ascii >= 97 and ascii <= 122 then -- lowercase
    return ascii - 96
  end -- uppercase
  return ascii - 38
end

local function part1(data)
  local totalPriority = 0
  for _, line in ipairs(data) do
    local lh, rh = line:sub(1, #line / 2), line:sub((#line / 2) + 1)
    local set = {}
    for c in rh:gmatch(".") do
      set[c] = true
    end
    for c in lh:gmatch(".") do
      if set[c] then
        totalPriority = totalPriority + getPriority(c)
        break
      end
    end
  end
  return totalPriority
end

local function part2(data, groupNumber)
  local totalPriority = 0
  for i = 1, #data, groupNumber do
    local countChar = {}
    for j = 0, groupNumber - 1 do
      local charSet = {}
      for c in data[i + j]:gmatch(".") do
        charSet[c] = true
      end
      for c in pairs(charSet) do
        countChar[c] = (countChar[c] or 0) + 1
        if countChar[c] == groupNumber then
          totalPriority = totalPriority + getPriority(c)
          break -- This will break the outer loop as well, bcs is the last iteration
        end
      end
    end
  end
  return totalPriority
end

local data = getData("input.txt")
print(part1(data))
print(part2(data, 3))

File: rust/day3/src/main.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::str::Lines;

fn get_priority(n: &u8) -> i32 {
    if (97..123).contains(n) {
        (n - 96) as i32
    } else {
        (n - 38) as i32
    }
}

fn part1(data: Lines) -> i32 {
    data.map(|l| l.as_bytes().split_at(l.len() / 2))
        .filter_map(|t| {
            let mut left = HashSet::<_>::from_iter(t.0);
            t.1.into_iter().find_map(|c| left.take(c))
        })
        .map(get_priority)
        .sum()
}

fn part2(data: Lines, group_number: usize) -> i32 {
    let v = data.map(|l| l.as_bytes()).collect::<Vec<_>>();
    let mut total_priority = 0;
    for group in v.chunks(group_number) {
        let mut counter_record = HashMap::new();
        for rucksack in group {
            for c in HashSet::<_>::from_iter(*rucksack) {
                let entry = counter_record
                    .entry(*c)
                    .and_modify(|counter| *counter += 1)
                    .or_insert(1);
                if *entry == group_number {
                    total_priority += get_priority(c);
                    break; // This also break the outter loop, bcs this only happens in the last rucksack
                }
            }
        }
    }
    total_priority
}

fn main() {
    let input = fs::read_to_string("input.txt").expect("Couldn't open file");
    println!(
        "Part1: {} Part2: {}",
        part1(input.lines()),
        part2(input.lines(), 3)
    );
}
